### ISSUE TITLE
BUG: Memory leak in buffer interface with `NPY_RELAXED_STRIDES_CHECKING`

### ISSUE THREAD

**Issue by peterbell10** [numpy/numpy#16934]
### Reproducing code example:
The following cython code leaks memory when called in a loop:

```python
import numpy as np
from cpython.buffer cimport *

a = np.random.rand(10)[:, None]

def memory_leak():
    cdef Py_buffer buf
    PyObject_GetBuffer(a, &buf, PyBUF_STRIDED)
    PyBuffer_Release(&buf)
```

The key factor seems to be the zero strided axis in `a`.

### Numpy/Python version information:
```
1.17.2 3.7.7 (default, May  7 2020, 21:25:33) 
[GCC 7.3.0]
```

**Comment by mattip** [comment]
Can you try with 1.19.1 if not with HEAD?

**Comment by peterbell10** [comment]
Tried on c3a887e55e0a14e3a31460e7a79e5f7f965fea68, the magnitude of the leak is slightly smaller but memory consumption still grows with each call.

**Comment by seberg** [comment]
Ouch, yeah, now that you mention it, I remember noticing that there may be a leak here in some cases.  The issue is that we clean up strides for the exported buffer so that buffer exporters don't have to be aware orf relaxed strides.  The issue is that this code happens after checking the buffer "cache" and not before.

**Comment by seberg** [comment]
I have to think about it a bit, the fix itself is pretty straight forward, but for some (rare) arrays, we have to check if the buffer is exported a second time with a different contiguity requested (i.e. first with F, then with C contiguity), since it would be strange if we change the strides of a buffer that may still be in use.

**Comment by seberg** [comment]
@peterbell10 how important is this to you? The fix is currently not backported to 1.19.x, and since 1.20 is almost out we are a bit in the balance of whether to backport it at all.  So if you say that it would help you to backport, that would probably be the tipping point for backporting.
